gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixJobLauncher.java

```diff
@@ -0,0 +1,273 @@
+            addWorkUnit(innerWorkUnit, stateSerDeRunner, taskConfigMap, jobName, jobId);
+          }
+        } else {
+          addWorkUnit(workUnit, stateSerDeRunner, taskConfigMap, jobName, jobId);
+        }
+      }
+    } catch (Throwable t) {
+      throw closer.rethrow(t);
+    } finally {
+      closer.close();
+    }
+
+    JobConfig.Builder jobConfigBuilder = new JobConfig.Builder();
+    jobConfigBuilder.addTaskConfigMap(taskConfigMap).setCommand(GobblinWorkUnitRunner.GOBBLIN_TASK_FACTORY_NAME);
+
+    return jobConfigBuilder;
+  }
+
+  /**
+   * Submit a job to run.
+   */
+  private void submitJobToHelix(String jobName, String jobId, JobConfig.Builder jobConfigBuilder) throws Exception {
+    // Create one queue for each job with the job name being the queue name
+    JobQueue jobQueue = new JobQueue.Builder(jobName).build();
+    try {
+      this.helixTaskDriver.createQueue(jobQueue);
+    } catch (IllegalArgumentException iae) {
+      LOGGER.warn(String.format("Job queue %s already exists", jobQueue.getName()));
+    }
+
+    // Put the job into the queue
+    this.helixTaskDriver.enqueueJob(jobName, jobId, jobConfigBuilder);
+  }
+
+  /**
+   * Add a single {@link WorkUnit} (flattened).
+   */
+  private void addWorkUnit(WorkUnit workUnit, ParallelRunner stateSerDeRunner,
+      Map<String, TaskConfig> taskConfigMap, String jobName, String jobId) throws IOException {
+    String workUnitFilePath = persistWorkUnit(new Path(this.inputWorkUnitDir, jobId), workUnit, stateSerDeRunner);
+
+    Map<String, String> rawConfigMap = Maps.newHashMap();
+    rawConfigMap.put(ConfigurationConstants.WORK_UNIT_FILE_PATH, workUnitFilePath);
+    rawConfigMap.put(ConfigurationKeys.JOB_NAME_KEY, jobName);
+    rawConfigMap.put(ConfigurationKeys.JOB_ID_KEY, jobId);
+    rawConfigMap.put(ConfigurationKeys.TASK_ID_KEY, workUnit.getId());
+    rawConfigMap.put("TASK_SUCCESS_OPTIONAL", "true");
+
+    LOGGER.info("Adding WorkUnit " + workUnit.getId());
+    taskConfigMap.put(workUnit.getId(), TaskConfig.from(rawConfigMap));
+  }
+
+  /**
+   * Persist a single {@link WorkUnit} (flattened) to a file.
+   */
+  private String persistWorkUnit(Path workUnitFileDir, WorkUnit workUnit, ParallelRunner stateSerDeRunner)
+      throws IOException {
+    Path workUnitFile = new Path(workUnitFileDir, workUnit.getId() + WORK_UNIT_FILE_EXTENSION);
+    stateSerDeRunner.serializeToFile(workUnit, workUnitFile);
+    return workUnitFile.toString();
+  }
+
+  private void waitForJobCompletion() throws InterruptedException {
+    while (true) {
+      WorkflowContext workflowContext = TaskUtil.getWorkflowContext(this.helixManager, this.helixQueueName);
+      if (workflowContext != null) {
+        org.apache.helix.task.TaskState jobState = workflowContext.getJobState(this.jobResourceName);
+        if (jobState != null && jobState == org.apache.helix.task.TaskState.COMPLETED) {
```

Comment:
If `jobState` is null then `jobState == org.apache.helix.task.TaskState.COMPLETED` should return false.